Src/FluidCaching/Index.cs

```diff
@@ -111,16 +101,11 @@ public int RebuildIndex()
         {
             lock (lifespanManager)
             {
-                return RWLock.GetWriteLock(_lock, LockTimeout, delegate
-                {
-                    index.Clear();
-                    foreach (INode<T> item in lifespanManager)
-                    {
-                        AddItem(item);
-                    }
-
-                    return index.Count;
-                });
+                // Create a new ConcurrentDictionary, this way there is no need for locking the index itself
+                var keyValues = lifespanManager.Select(item => new KeyValuePair<TKey, WeakReference<INode<T>>>(_getKey(item.Value), new WeakReference<INode<T>>(item)));
```

Comment:
We would need a need a ToConcurentDictionary() for that, which is not available in the framework. Implementing it here for single use seems a bit overkill.

I do agree this code is bit complex, mainly due to all the nested generics (imagine if i had not used var here :-) something like IEnumerable<KeyValuePair<Tkey, WeakReference<INode<T>>>